core/src/main/scala/org/apache/spark/scheduler/TaskLocation.scala

```diff
@@ -64,14 +64,14 @@ private[spark] object TaskLocation {
 
   /**
    * Create a TaskLocation from a string returned by getPreferredLocations.
-   * These strings have the form [hostname] or hdfs_cache_[hostname], depending on whether the
-   * location is cached.
+   * These strings have the form [executorLocationTag][hostname][executorid], [hostname], or
+   * hdfs_cache_[hostname], depending on whether the location is cached.
    */
   def apply(str: String): TaskLocation = {
     val hstr = str.stripPrefix(inMemoryLocationTag)
     if (hstr.equals(str)) {
       if (str.startsWith(executorLocationTag)) {
-        val splits = str.split("_")
+        val splits = str.split("_", 3)
```

Comment:
OK, probably even more robust to just use stripPrefix here? this whole method could be a little clearer anyway